### ISSUE TITLE
Ceph operator iterating over all Nodes instead of just OSD placement valid Nodes

### ISSUE THREAD

**Issue by galexrt** [rook/rook#6111]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

The operator is iterating over all Nodes on startup and not just the Nodes which are valid for OSDs, see https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/watcher.go#L81.

It is operating over all Nodes because Kubernetes sends all objects as "Created" when a watch is created.

**Expected behavior:**

The operator to only iterate over the Nodes on startup that are valid for OSDs placement.

**How to reproduce it (minimal and precise):**

1. Not having a `all` placement which keeps Rook Ceph on the nodes with OSDs.
	* (Optional) Remove `all` placement from CephCluster object.
1. (Re-)Start the operator.
1. Check the operator logs, it will begin logging out an  "error" per node (without OSDs) but continue normally (error is correctly handled) with the iteration and in the end, the Nodes watch loop.

**Environment**:
* Rook version (use `rook version` inside of a Rook Pod): v1.4.1
* Kubernetes version (use `kubectl version`): `v1.18.6`
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Bare metal

**Comment by travisn** [comment]
The purpose of watching the nodes is purely to trigger a reconcile when a new node is added so we can configure new OSD(s) on the new node(s). At startup of the operator a full reconcile is already triggered so we should skip these events if at all possible at startup. There really isn't a reason to evaluate the nodes in that case, not even just the nodes matching the OSD placement.

**Comment by galexrt** [comment]
@travisn What about introducing two changes to the logic around node watching:

1. Only watch Nodes based on OSD placement. This could be done as a quick fix for big K8S clusters with just a few OSDs (e.g., K8S cluster has 50+ nodes and only 5 are OSDs).
2. Ignore node events during first reconcile and / or while reconcile is running.
    * When ignoring node events while reconcile is running is that some "added" nodes might not be picked up because the reconcile is "beyond" the OSD creation process "already".

What do you think?

**Comment by travisn** [comment]
> @travisn @leseb What about introducing two changes to the logic around node watching:
> 
> 1. Only watch Nodes based on OSD placement. This could be done as a quick fix for big K8S clusters with just a few OSDs (e.g., K8S cluster has 50+ nodes and only 5 are OSDs).

Yes, makes sense to limit the watch based on the osd/all merged placement. A couple more considerations:
- If the osd placement isn't set, that implies watching all nodes
- There is no need to watch for new nodes if `useAllNodes: false`

> 2. Ignore node events during first reconcile and / or while reconcile is running.
>    
>    * When ignoring node events while reconcile is running is that some "added" nodes might not be picked up because the reconcile is "beyond" the OSD creation process "already".

Agreed, let's ignore the events during the first reconcile. After the first reconcile is done, then we stop ignoring the add node events. You're right that it is possible that events will be ignored if added during that first reconcile. This seems uncommon enough to be acceptable.

**Comment by galexrt** [comment]
Reopening as first part has been and second part will follow.
